### PR TITLE
Check for null connections in resource layer before using them

### LINKED ISSUE #151: Check we have a connection before trying to use it
It is possible to pass a null to a resource for the connection. This is allowed as a resource maybe being created before it knows where it is being pushed to. We need to ensure that we have a connection before we try to use it.

### PR DESCRIPTION
**Pull request by pnickoll**
fixes #151

### CODE REVIEW

File: client-lib/src/main/java/com/ibm/ws/repository/exceptions/RepositoryResourceNoConnectionException.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+* limitations under the License.
+*******************************************************************************/
+package com.ibm.ws.repository.exceptions;
+
+/**
+ * This is thrown when an operation is attempted on a resource that requires a connection and the resource does not have a connection.
+ */
+public class RepositoryResourceNoConnectionException extends RepositoryResourceException {
+
+    private static final long serialVersionUID = 8976559783458635686L;
```
Comment (idlewis) [inline client-lib/src/main/java/com/ibm/ws/repository/exceptions/RepositoryResourceNoConnectionException.java]: Do we need a real version id? I kind of prefer to have the default one here, as it indicates that no one has actually thought about whether this is serializable or not.

File: client-lib/src/main/java/com/ibm/ws/repository/strategies/writeable/BaseStrategy.java [hunk 2]
```diff
@@ -93,10 +94,11 @@ protected State calculateTargetState(RepositoryResourceImpl matchingResource) {
 
     /**
      * Delegates to the resource.findMatchingResource, see {@link RepositoryResourceImpl#findMatchingResource()}
+     * 
+     * @throws RepositoryResourceNoConnectionException
```
Comment (idlewis) [inline client-lib/src/main/java/com/ibm/ws/repository/strategies/writeable/BaseStrategy.java]: Would it be better to add all exceptions to the javadoc? Or none, I'd be happy with either

File: client-lib/src/main/java/com/ibm/ws/repository/strategies/writeable/UpdateInPlaceStrategy.java [hunk 3]
```diff
@@ -149,9 +148,8 @@ public void uploadAsset(RepositoryResourceImpl resource, List<RepositoryResource
      * @throws RepositoryBadDataException
      * @throws RepositoryResourceCreationException
      */
-    public void uploadAttachment(RepositoryResourceImpl resource, AttachmentResourceImpl attachment, RepositoryResourceImpl matchingResource)
-                    throws RepositoryResourceCreationException, RepositoryBadDataException, RepositoryResourceUpdateException,
-                    RepositoryBackendException, RepositoryResourceException {
+    public void uploadAttachment(RepositoryResourceImpl resource, AttachmentResourceImpl attachment,
```
Comment (idlewis) [inline client-lib/src/main/java/com/ibm/ws/repository/strategies/writeable/UpdateInPlaceStrategy.java:151]: The exception isn't added to the javadoc here. I suspect there are other similar issues elsewhere. Don't know if the best way to fix is to remove the other exceptions or add the new one

File: client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java [hunk 4]
```diff
@@ -161,10 +162,16 @@ private RepositoryWriteableClient getWritableClient() throws RepositoryOperation
         if (_client instanceof RepositoryWriteableClient) {
             return (RepositoryWriteableClient) _client;
         } else {
-            throw new RepositoryOperationNotSupportedException("The backend does not support write operations", _repoConnection);
+            throw new RepositoryOperationNotSupportedException("The backend does not support write operations or no connection has been specified", getRepositoryConnection());
```
Comment (idlewis) [inline client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java:165]: It doesn't look like the new message matches up with the code. I can't see a check for whether the connection has been specified or not.

File: client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java [hunk 5]
```diff
@@ -337,6 +344,14 @@ public RepositoryConnection getRepositoryConnection() {
         return _repoConnection;
     }
 
+    public RepositoryConnection getAndCheckRepositoryConnection() throws RepositoryResourceNoConnectionException {
+        if (_repoConnection == null) {
+            // In order to have an ID then we should have already been stored so we should have a connection but just in case...
+            throw new RepositoryResourceNoConnectionException("No connection has been specified for the resource so unable to getAttachments", getId());
```
Comment (idlewis) [inline client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java]: the exception message looks wrong. Why is it talking about attachments?

File: client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java [hunk 6]
```diff
@@ -1677,11 +1685,11 @@ public InputStream getInputStream() throws RepositoryBackendException, Repositor
             try {
                 return _client.getAttachment(_asset, _attachment);
             } catch (IOException e) {
-                throw new RepositoryBackendIOException("Failed to get read attachment", e, _repoConnection);
+                throw new RepositoryBackendIOException("Failed to get read attachment", e, getRepositoryConnection());
```
Comment (idlewis) [inline client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java:1696]: This looks odd. In the constructor of an exception we are calling a method that may well throw an exception. That would be hugely confusing if it happened

File: client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java (hunk 4 above)
Comment (pnickoll) [reply client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java:165]: If a connection has not been specified a _client won't have been created, so _client will be null and therefore not an instance of RepositoryWritabelClent

File: client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java (hunk 5 above)
Comment (pnickoll) [reply client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java]: Oops cut and paste error, will fix

File: client-lib/src/main/java/com/ibm/ws/repository/strategies/writeable/UpdateInPlaceStrategy.java (hunk 3 above)
Comment (pnickoll) [reply client-lib/src/main/java/com/ibm/ws/repository/strategies/writeable/UpdateInPlaceStrategy.java:151]: I'll look at everywhere I added the new exception and update missing javadoc

File: client-lib/src/main/java/com/ibm/ws/repository/strategies/writeable/BaseStrategy.java (hunk 2 above)
Comment (pnickoll) [reply client-lib/src/main/java/com/ibm/ws/repository/strategies/writeable/BaseStrategy.java]: Agree, I'll take a look

File: client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java (hunk 6 above)
Comment (pnickoll) [reply client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java:1696]: getRepositoryConnection doesn't throw an exception, getAndCheckRepositoryConnection does

File: client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java (hunk 4 above)
Comment (idlewis) [reply client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java:165]: If _client is null, then your instanceof check will NPE?

Comment (idlewis) [reply client-lib/src/main/java/com/ibm/ws/repository/resources/internal/RepositoryResourceImpl.java:165]: ah, no it won't. I need to read the java language spec more carefully

File: client-lib/src/main/java/com/ibm/ws/repository/exceptions/RepositoryResourceNoConnectionException.java (hunk 1 above)
Comment (pnickoll) [reply client-lib/src/main/java/com/ibm/ws/repository/exceptions/RepositoryResourceNoConnectionException.java]: Yeah that sounds more sensible, will change

**Review by idlewis — APPROVED** [review]

**Pull request merged**